### PR TITLE
Signed Extensions

### PR DESCRIPTION
**Pull request by hannes**
DuckDB can `INSTALL` various extensions from our centrally managed extensions server. Those downloads use HTTP and not HTTP*S* because the latter would require a dependency on OpenSSL or something equally heavy. However, this creates a potential security issue where an attacker could use variuos spoofing methods to redirect the download to retrieve a malicious binary which is then dlopen()ed and effectively executed. We are not aware of any such attacks occurring in the wild, but they are concievable.

To address this, we are adding cryptographic signatures to the extension binaries. Signatures consist of hashes of the binary that are encrypted with a 2048 bit RSA private key. On extension `LOAD`, we verify these signatures. We decrypt the hash in the signature using the public key and recompute the hash of the binary. Only if they match loading will proceed. For the crytographic parts, we ship a small portion of the excellent [Mbed-TLS library](https://github.com/Mbed-TLS/mbedtls) from ARM. `extension_helper.cpp` contains a set of 20 RSA public keys that are all tried to verify an extension signature. The corresponding private keys are held by the DuckDB maintainers, Hannes and Mark.

For development, it is possible to disable extension signature checking by configuration. A new configuration option is added, `allow_unsigned_extensions`. The DuckDB command line interface (shell) gains a new command line parameter, `-unsigned`. 

If you would like your extension to be built, signed and shipped from the main DuckDB CI and extension repository, please get in touch. We have created a method to include out-of-tree extensions.

### CODE REVIEW

**Comment by hannes** [comment]
CC @zenazn

**Review by samansmink — APPROVED** [review]
Very nice! lgtm

File: src/main/extension/extension_load.cpp [hunk 1]
```diff
@@ -37,11 +38,44 @@ void ExtensionHelper::LoadExternalExtension(DatabaseInstance &db, const string &
 	}
 
 	if (!fs.FileExists(filename)) {
-		throw IOException("File \"%s\" not found", filename);
+		throw IOException("Extension \"%s\" not found", filename);
+	}
+	{
+		auto handle = fs.OpenFile(filename, FileFlags::FILE_FLAGS_READ);
+
+		// signature is the last 265 bytes of the file
```
Comment (zenazn) [inline src/main/extension/extension_load.cpp:46]: 256

**Comment by zenazn** [comment]
thank you!

**Comment by bucweat** [comment]
Hi! Catching up on progress made over the past couple of months :-)

I would like to use `sqlite_scanner` which I compiled locally via the ODBC interface in Windows. I have the ODBC driver installed and it works well (thanks @tiagokepe). The local version of `sqlite_scanner` is also working in the CLI when the `-unsigned` option is used. So I think I have all the parts I need to make this work.

When I try to load the extension via ODBC using `LOAD 'sqlite_scanner';` I get the error `[extension] could not be loaded because its signature is either missing or invalid and unsigned extensions are disabled by configuration (allow_unsigned_extensions)`.  The error is not unexpected given the documented security requirements for extensions. 

My problem is I'm having trouble figuring out how to set the value of `allow_unsigned_extensions`. I tried `PRAGMA allow_unsigned_extensions='true';` but get the error `Error: Cannot change allow_unsigned_extensions setting while database is running`. Ditto with the CLI which I'm assuming is why there is a `-unsigned` command line option. Given how ODBC works, you typically pass options via the connection string, but from a quick code search I'm guessing that option is not currently implemented in the ODBC driver. Am I wrong? Is there another way to set `allow_unsigned_extensions` via ODBC that I'm missing?

**Comment by Mytherin** [comment]
Modifying `allow_unsigned_extensions` at run-time is prohibited for safety reasons, the flag needs to be set at database start-up. We have key/value options that allow setting this in various clients, but it is possible that this hasn't been ported to the ODBC client yet.

**Pull request merged**